### ISSUE TITLE
Random panics at closing file descriptors

### ISSUE THREAD

**Issue by mcseemk** [tokio-rs/tokio-uring#35]
I observe random panics when running the following code:

```rust
async fn compute_tokio_uring() {
    use tokio::task::JoinHandle;
    use tokio_uring::fs::File;

    let handles: Vec<JoinHandle<_>> = (0..1_000)
        .map(|_| {
            tokio_uring::spawn(async {
                let buffer = vec![0; 10];
                let dev_urandom = File::open("/dev/urandom").await.unwrap();
                let (res, buffer) = dev_urandom.read_at(buffer, 0).await;
                let _read_n = res.unwrap();
                let _ = dev_urandom.close().await;
                let dev_null = File::open("/dev/null").await.unwrap();
                let _ = dev_null.write_at(buffer, 0).await;
                let _ = dev_null.close().await;
            })
        })
        .collect();
    for handle in handles {
        handle.await.unwrap();
    }
}

fn main() {
    tokio_uring::start(async {
        for _ in 0..10 {
            compute_tokio_uring().await;
        }
        for _ in 0..10 {
            let before = std::time::Instant::now();
            for _ in 0..1_000 {
                compute_tokio_uring().await;
            }
            let elapsed = before.elapsed();
            println!(
                "[tokio_uring] {:?} total, {:?} avg per iteration",
                elapsed,
                elapsed / 1_000
            );
        }
    })
}
```

Typical output:

```
[tokio_uring] 3.140604502s total, 3.140604ms avg per iteration
[tokio_uring] 3.138823337s total, 3.138823ms avg per iteration
[tokio_uring] 3.182483812s total, 3.182483ms avg per iteration
thread 'main' panicked at 'already borrowed: BorrowMutError', /home/mcs/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-uring-0.1.0/src/driver/op.rs:151:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'already borrowed: BorrowMutError', /home/mcs/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-uring-0.1.0/src/driver/op.rs:151:37
stack backtrace:
   0:     0x560a89614b66 - std::backtrace_rs::backtrace::libunwind::trace::h0f5cd2ee8b0d7274
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x560a89614b66 - std::backtrace_rs::backtrace::trace_unsynchronized::h06905b5aeda069a1
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x560a89614b66 - std::sys_common::backtrace::_print_fmt::h4fe4c7c875072f30
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x560a89614b66 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcc0746f004a9b7ef
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x560a895fdcec - core::fmt::write::h9a6d9c74526a6c1b
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/core/src/fmt/mod.rs:1115:17
   5:     0x560a89612972 - std::io::Write::write_fmt::h23dab4cc9ce72ee2
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/io/mod.rs:1665:15
   6:     0x560a89613b58 - std::sys_common::backtrace::_print::h173dc702502d65d2
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x560a89613b58 - std::sys_common::backtrace::print::h61bd27c4742ba817
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x560a89613b58 - std::panicking::default_hook::{{closure}}::hcaae87f0495ae613
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/panicking.rs:208:50
   9:     0x560a89612584 - std::panicking::default_hook::h0538e728ee080db0
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/panicking.rs:225:9
  10:     0x560a89612584 - std::panicking::rust_panic_with_hook::h3039e236b6ca482c
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/panicking.rs:622:17
  11:     0x560a8962d7f8 - std::panicking::begin_panic_handler::{{closure}}::h884fbab544ffd91c
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/panicking.rs:519:13
  12:     0x560a8962d76e - std::sys_common::backtrace::__rust_end_short_backtrace::hdaf2e18ba3d91210
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x560a8962d71d - rust_begin_unwind
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/std/src/panicking.rs:515:5
  14:     0x560a895fceb0 - core::panicking::panic_fmt::hcf5f6d96e1dd7099
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/core/src/panicking.rs:92:14
  15:     0x560a895fe1e2 - core::result::unwrap_failed::he898b02f57993c42
                               at /rustc/32c9b7b091534f6d80e7e85da0cd425acb6c9a79/library/core/src/result.rs:1599:5
  16:     0x560a8964d1a7 - core::ptr::drop_in_place<tokio_uring::driver::op::Op<tokio_uring::driver::close::Close>>::hfcf27dfadca700b2
  17:     0x560a8964cdde - tokio_uring::driver::shared_fd::Inner::submit_close_op::h3f8bddef42430763
  18:     0x560a895ef99e - core::ptr::drop_in_place<tokio_uring::fs::file::File>::h362afd75267e2033
  19:     0x560a895f4a25 - tokio::runtime::task::raw::poll::h085f63b72b0ae836
  20:     0x560a895f8583 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h3c0c4df751b39350
  21:     0x560a895fac6a - probing::main::h251d5f151cb99f6d
  22:     0x560a895ec1b3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h7ba9bcd7fa98fc5f
  23:     0x560a895fc247 - main
  24:     0x7f946e667565 - __libc_start_main
  25:     0x560a895ec08e - _start
  26:                0x0 - <unknown>
thread panicked while panicking. aborting.
```

After panic while panicking such process hangs forever, even kill -9 cannot terminate it.

Sometimes the result is slightly different:
```
...
thread panicked while panicking. aborting.
Illegal instruction (core dumped)
```

Tested on the following systems:

    Linux 5.11.0-25-generic #27-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux
    Linux 5.8.0-1038-aws #40~20.04.1-Ubuntu SMP Thu Jun 17 13:25:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Rust version

    rustc 1.55.0-nightly (027187094 2021-07-22)

and also on the previous rust nightly (2021-07-21)

**Comment by JakubKoralewski** [comment]
Takes really long to run, so couldn't replicate. I'm getting a 10x slower execution time:
```
tokio_uring] 40.370097994s total, 40.370097ms avg per iteration
```

```
Linux 5.11.0-7620-generic #21~1626191760~21.04~55de9c3-Ubuntu SMP Tue Jul 20 22:18:55 UTC  x86_64 GNU/Linux
```

same speed on both stable 1.53 and `1.55.0-nightly (4e143fd13 2021-07-20)
`

**Comment by mcseemk** [comment]
Looks like you used a debug build, hence slower. I used release since wanted to benchmark against vanilla tokio.
Anyway, looks like some race condition to me.

**Comment by songzhi** [comment]
In my platform, seems the problem is from `/dev/urandom`. When I run the code, it sometimes paused forever without any panic. When I use a regular path, it works fine and was 2.x ms each iteration slower than yours, used the release build. 

My toolchain and platform is:
```
nightly-x86_64-unknown-linux-gnu (default)
rustc 1.55.0-nightly (32c9b7b09 2021-07-21)

Linux 5.13.4-1-MANJARO #1 SMP PREEMPT Tue Jul 20 21:25:32 UTC 2021 x86_64 GNU/Linux
```

**Comment by mcseemk** [comment]
On my PC the release build reading from regular text file crashes immediately with the same error, but I was not able to reproduce the crash with the debug build though.

**Comment by carllerche** [comment]
Thanks for the report. I will investigate.

**Comment by carllerche** [comment]
I think I know what is happening, but I am not sure how to reproduce reliably. I pushed a possible fix to #39, could someone give it a shot? I will try to keep trying to reproduce myself.
